src/com/facebook/buck/go/GoLinkable.java

```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright 2015-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.go;
+
+import com.facebook.buck.model.BuildTargets;
+import com.facebook.buck.rules.BuildRule;
+import com.facebook.buck.rules.Tool;
+import com.facebook.buck.rules.AbstractBuildRule;
+import com.facebook.buck.rules.AddToRuleKey;
+import com.facebook.buck.rules.BuildContext;
+import com.facebook.buck.rules.BuildRuleParams;
+import com.facebook.buck.rules.BuildableContext;
+import com.facebook.buck.rules.SourcePath;
+import com.facebook.buck.rules.SourcePathResolver;
+import com.facebook.buck.step.Step;
+import com.facebook.buck.step.fs.MakeCleanDirectoryStep;
+import com.facebook.buck.step.fs.MkdirStep;
+import com.facebook.buck.step.fs.SymlinkTreeStep;
+import com.facebook.buck.util.HumanReadableException;
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.io.Files;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+abstract class GoLinkable extends AbstractBuildRule {
+  @AddToRuleKey
+  private final Tool compiler;
+  @AddToRuleKey(stringify = true)
+  private final Path packageName;
+  @AddToRuleKey
+  private final ImmutableSet<SourcePath> srcs;
+  @AddToRuleKey
+  private final ImmutableList<String> flags;
+
+  private final Path scratchDir;
+  private final Path output;
+
+  GoLinkable(
+      BuildRuleParams params,
+      SourcePathResolver resolver,
+      Optional<Path> packageName,
+      ImmutableSet<SourcePath> srcs,
+      ImmutableList<String> flags,
+      Path output,
+      Tool compiler) {
+    super(params, resolver);
+    this.srcs = srcs;
+    this.packageName = packageName.or(getBuildTarget().getBasePath());
+    this.flags = flags;
+    this.output = output;
+    this.compiler = compiler;
+    this.scratchDir = BuildTargets.getScratchPath(getBuildTarget(), "__%s__compile");
+  }
+
+  @Override
+  public ImmutableList<Step> getBuildSteps(
+      BuildContext context,
+      BuildableContext buildableContext) {
+    ImmutableMap.Builder<Path, Path> symlinkDeps = ImmutableMap.builder();
+    for (BuildRule buildRule : getDeps()) {
```

Comment:
As above this kind of logic is usually in the description.

Since this would be shared between the binary and library descriptions you can extract it into a static method in a `GoDescriptions` class.